File: go/verify/verifier.go
package verify

import (
	"crypto/ecdsa"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/Truelayer/truelayer-signing/go/crypto"
	"github.com/Truelayer/truelayer-signing/go/errors"
	"github.com/Truelayer/truelayer-signing/go/jws"
	"github.com/Truelayer/truelayer-signing/go/sign"
)

// Build to verify a request against a "Tl-Signature" header.
type Verifier struct {
	publicKey       []byte
	jwks            []byte
	body            []byte
	method          string
	path            string
	headers         map[string][]byte
	requiredHeaders []string
}

// Truelayer signature data
type TlSignature struct {
	JwsHeader    *jws.JwsHeader
	HeaderBase64 string
	Signature    []byte
}

func NewVerifier(publicKeyPem []byte) *Verifier {
	verifier := Verifier{}
	verifier.publicKey = publicKeyPem
	verifier.method = ""
	verifier.path = ""
	verifier.body = []byte("")
	verifier.headers = make(map[string][]byte)
	return &verifier
}

func NewVerifierWithJwks(jwks []byte) *Verifier {
	verifier := Verifier{}
	verifier.jwks = jwks
	verifier.method = ""
	verifier.path = ""
	verifier.body = []byte("")
	verifier.headers = make(map[string][]byte)
	return &verifier
}

// Body adds the full received request body.
func (v *Verifier) Body(body []byte) *Verifier {
	v.body = body
	return v
}

// Method adds the request method, e.g. "POST".
func (v *Verifier) Method(method string) *Verifier {
	v.method = method
	return v
}

// Path adds the request path, e.g. "/payouts".
func (v *Verifier) Path(path string) *Verifier {
	v.path = path
	return v
}

// Header adds a header name & value.
// May be called multiple times to add multiple different headers.
//
// All request headers may be added here, any headers not mentioned
// in the jws signature header will be ignored unless required
// using "RequiredHeader".
func (v *Verifier) Header(name string, value []byte) *Verifier {
	v.headers[strings.ToLower(name)] = value
	return v
}

// Headers appends multiple header name & value.
//
// Warning: Only a single value per header name is supported.
func (v *Verifier) Headers(headers map[string][]byte) *Verifier {
	for name, value := range headers {
		v.Header(name, value)
	}
	return v
}

// RequireHeader specifies a header name that must be included in the "Tl-Signature".
// May be called multiple times to add multiple required headers.
//
// Signatures missing these will fail verification.
func (v *Verifier) RequireHeader(name string) *Verifier {
	v.requiredHeaders = append(v.requiredHeaders, name)
	return v
}

// Verify verifies the given "Tl-Signature" header value.
//
// Supports v1 (body-only) & v2 full request signatures.
//
// Returns error if verification fails.
func (v *Verifier) Verify(tlSignature string) error {
	if !strings.HasPrefix(v.path, "/") {
		return errors.NewInvalidArgumentError("path must start with '/'")
	}

	tlSignatureData, err := ParseTlSignature(tlSignature)
	if err != nil {
		return errors.NewJwsError(fmt.Sprintf("signature parsing failed: %v", err))
	}
	jwsHeader := tlSignatureData.JwsHeader

	var publicKey *ecdsa.PublicKey
	if v.publicKey == nil {
		if v.jwks == nil {
			return errors.NewInvalidKeyError("no public key nor jwks supplied: verification is not possible")
		} else {
			publicKey, err = crypto.FindAndParseEcJwk([]byte(jwsHeader.Kid), v.jwks)
			if err != nil {
				return errors.NewInvalidKeyError(fmt.Sprintf("jwk find and parse failed: %v", err))
			}
		}
	} else {
		publicKey, err = crypto.ParseEcPublicKey(v.publicKey)
		if err != nil {
			return errors.NewInvalidKeyError(fmt.Sprintf("public key parsing failed: %v", err))
		}
	}

	if jwsHeader.Alg != "ES512" {
		return errors.NewJwsError(fmt.Sprintf("unexpected header alg: %s", jwsHeader.Alg))
	}

	if jwsHeader.TlVersion == "" || jwsHeader.TlVersion == "1" {
		return errors.NewJwsError("v1 signature not allowed")
	}

	// check and order all required headers
	orderedHeaders, err := jwsHeader.FilterHeaders(v.headers)
	if err != nil {
		return errors.NewJwsError(fmt.Sprintf("headers filtering failed: %v", err))
	}
	// fail if signature is missing a required header
	for _, header := range v.requiredHeaders {
		_, exists := orderedHeaders.Get(strings.ToLower(header))
		if !exists {
			return errors.NewJwsError(fmt.Sprintf("signature is missing required header: %s", header))
		}
	}

	// reconstruct the payload as it would have been signed
	signingPayload := sign.BuildV2SigningPayload(v.method, v.path, orderedHeaders, v.body, false)
	payload := fmt.Sprintf("%s.%s", tlSignatureData.HeaderBase64, base64.RawURLEncoding.EncodeToString(signingPayload))
	err = crypto.VerifyES512(publicKey, []byte(payload), tlSignatureData.Signature)
	if err != nil {
		// try again with/without a trailing slash (#80)
		newPath, addPathTrailingSlash := v.handleTrailingSlashRetry()
		signingPayload := sign.BuildV2SigningPayload(v.method, newPath, orderedHeaders, v.body, addPathTrailingSlash)
		payload := fmt.Sprintf("%s.%s", tlSignatureData.HeaderBase64, base64.RawURLEncoding.EncodeToString(signingPayload))
		if retryErr := crypto.VerifyES512(publicKey, []byte(payload), tlSignatureData.Signature); retryErr != nil {
			// use original error if both fail
			return errors.NewJwsError(fmt.Sprintf("signature verification failed: %v", err))
		}
	}

	return nil
}

func (v *Verifier) handleTrailingSlashRetry() (string, bool) {
	if strings.HasSuffix(v.path, "/") {
		return strings.TrimSuffix(v.path, "/"), false
	} else {
		return v.path, true
	}
}

// ParseTlSignature parses a tl signature header value into (header, headerBase64, signature).
func ParseTlSignature(tlSignature string) (*TlSignature, error) {
	splits := strings.Split(tlSignature, "..")
	if len(splits) != 2 {
		return nil, fmt.Errorf("invalid signature format")
	}
	headerB64 := splits[0]
	signatureB64 := splits[1]
	header, err := base64.RawURLEncoding.DecodeString(headerB64)
	if err != nil {
		return nil, fmt.Errorf("jws header base64 decode failed: %v", err)
	}
	signature := make([]byte, base64.RawURLEncoding.DecodedLen(len(signatureB64)))
	length, err := base64.RawURLEncoding.Decode(signature, []byte(signatureB64))
	if err != nil {
		return nil, fmt.Errorf("signature base64 decode failed: %v", err)
	}
	signature = signature[:length]

	var jwsHeader jws.JwsHeader
	err = json.Unmarshal(header, &jwsHeader)
	if err != nil {
		return nil, fmt.Errorf("jws header json unmarshalling failed: %v", err)
	}

	return &TlSignature{
		JwsHeader:    &jwsHeader,
		HeaderBase64: headerB64,
		Signature:    signature,
	}, nil
}

File: ruby/lib/truelayer-signing/verifier.rb
# frozen_string_literal: true

module TrueLayerSigning
  class Verifier < JwsBase
    EXPECTED_EC_KEY_COORDS_LENGTH = 66

    attr_reader :required_headers, :key_type, :key_value

    def initialize(args)
      super

      @key_type = args[:key_type]
      @key_value = args[:key_value]
    end

    def verify(tl_signature)
      ensure_verifier_config!

      jws_header, jws_header_b64, signature_b64 = self.class.parse_tl_signature(tl_signature)

      validate_algorithm!(jws_header.alg)

      ordered_headers = jws_header.filter_headers(headers)
      normalised_headers = Utils.normalise_headers!(ordered_headers)

      validate_required_headers!(normalised_headers)

      verify_signature_flex(ordered_headers, jws_header, jws_header_b64, signature_b64)
    end

    def require_header(name)
      @required_headers ||= []
      @required_headers.push(name)

      self
    end

    def require_headers(names)
      @required_headers = names

      self
    end

    def self.parse_tl_signature(tl_signature)
      jws_header_b64, signature_b64 = tl_signature.split("..")

      raise(Error, "Invalid signature format") unless signature_b64

      begin
        jws_header_raw = Base64.urlsafe_decode64(jws_header_b64)
      rescue ArgumentError
        raise(Error, "Invalid base64 for header")
      else
        jws_header = JwsHeader.new(JSON.parse(jws_header_raw, symbolize_names: true))
      end

      [jws_header, jws_header_b64, signature_b64]
    end

    private

    def verify_signature_flex(ordered_headers, jws_header, jws_header_b64, signature_b64)
      full_signature = build_full_signature(ordered_headers, jws_header_b64, signature_b64)

      begin
        verify_signature(jws_header, full_signature)
      rescue JWT::VerificationError
        @path = path.end_with?("/") ? path[0...-1] : "#{path}/"
        full_signature = build_full_signature(ordered_headers, jws_header_b64, signature_b64)

        begin
          verify_signature(jws_header, full_signature)
        rescue JWT::VerificationError
          raise(Error, "Signature verification failed")
        end
      end
    end

    def build_full_signature(ordered_headers, jws_header_b64, signature_b64)
      payload_b64 = Base64.urlsafe_encode64(build_signing_payload(ordered_headers), padding: false)

      [jws_header_b64, payload_b64, signature_b64].join(".")
    end

    def verify_signature(jws_header, full_signature)
      case key_type
      when :pem
        public_key = OpenSSL::PKey.read(key_value)
      when :jwks
        public_key = retrieve_public_key(:jwks, key_value, jws_header)
      end

      jwt_options = {
        algorithm: TrueLayerSigning.algorithm,
        verify_expiration: false,
        verify_not_before: false
      }

      JWT.truelayer_decode(full_signature, public_key, jwt_options)
    end

    def retrieve_public_key(key_type, key_value, jws_header)
      case key_type
      when :pem
        OpenSSL::PKey.read(key_value)
      when :jwks
        jwks_hash = JSON.parse(key_value, symbolize_names: true)
        jwk = jwks_hash[:keys].find { |key| key[:kid] == jws_header.kid }

        raise(Error, "JWKS does not include given `kid` value") unless jwk

        JWT::JWK::EC.import(jwk).public_key
      else
        raise(Error, "Type of public key not recognised")
      end
    end

    def validate_required_headers!(headers)
      raise(Error, "Signature missing required header(s)") if required_headers&.any? do |key|
        !headers.key?(key.downcase)
      end
    end

    def validate_algorithm!(algorithm)
      raise(Error, "Unexpected `alg` header value") if algorithm != TrueLayerSigning.algorithm
    end

    def ensure_verifier_config!
      raise(Error, "Key value missing") unless key_value
    end
  end
end
